pytensor/tensor/math.py

```diff
@@ -2366,7 +2361,92 @@ def tensordot(a, b, axes=2):
     See the documentation of numpy.tensordot for more examples.
 
     """
-    return _tensordot_as_dot(a, b, axes, dot=dot, batched=False)
+    try:
+        iter(axes)
+    except Exception:
+        axes_a = list(range(-axes, 0))
+        axes_b = list(range(0, axes))
+    else:
+        axes_a, axes_b = axes
+    try:
+        na = len(axes_a)
+        axes_a = list(axes_a)
+    except TypeError:
+        axes_a = [axes_a]
+        na = 1
+    try:
+        nb = len(axes_b)
+        axes_b = list(axes_b)
+    except TypeError:
+        axes_b = [axes_b]
+        nb = 1
+
+    a = as_tensor_variable(a)
+    b = as_tensor_variable(b)
+    as_ = a.shape
+    bra = a.broadcastable
+    ats = a.type.shape
+    nda = a.ndim
+    bs = b.shape
+    brb = b.broadcastable
+    bts = b.type.shape
+    ndb = b.ndim
+    if na != nb:
+        raise ValueError(
+            "The number of axes supplied for tensordot must be equal for each tensor. "
+            f"Got {na} and {nb} respectively."
+        )
+    for k in range(na):
+        ax_a = axes_a[k]
+        ax_b = axes_b[k]
+        if ax_a < 0:
+            axes_a[k] += nda
+        if axes_a[k] < 0 or axes_a[k] >= nda:
+            raise ValueError(
+                f"Supplied axes {ax_a} for first input of tensordot is out of bounds. "
+                f"Input tensor has only ndim={nda}."
+            )
+        if ax_b < 0:
+            axes_b[k] += ndb
+        if axes_b[k] < 0 or axes_b[k] >= ndb:
+            raise ValueError(
+                f"Supplied axes {ax_b} for first input of tensordot is out of bounds. "
+                f"Input tensor has only ndim={ndb}."
+            )
+        if (bra[ax_a] != brb[ax_b]) or (
+            ats[ax_a] is not None and bts[ax_b] is not None and ats[ax_a] != bts[ax_b]
+        ):
+            raise ValueError(
+                "Input arrays have inconsistent broadcastable pattern or type shape along the axes "
+                "must be multiplied and summed with tensordot."
+            )
+        elif ats[ax_a] is None or bts[ax_b] is None:
+            a = Assert(
```

Comment:
Asserts are pretty bad for optimization, can we do without them?

Reply:
I try to do without them as much as possible but `tensordot` should fail to work if you ever try to do:

```python
a = tensor(shape=(3, 2))
b = tensor(shape=(2, 3))
c = tensordot(a, b, axes=[[0, 1], [0, 1]])
```

The tensors `a` and `b` have the same total number of elements, so `dot(a.flatten(), b.flatten())` is guaranteed to work (something similar to this flatten is what `tensordot` is doing behind the scenes). The problem is that `tensordot` **should raise an exception** because the `a.shape[0] != b.shape[0]` and `a.shape[1] != b.shape[1]`. That's why I only ever add an assert when the static shape information is not available to check whether `a.shape[i] == b.shape[i]` while making the node.

Reply:
Yeah that makes sense

Reply:
Micro-optimization idea: you can skip the check for the first unknown entry. This way if there's a single axis you don't need it (the dot op does that check internally) and it still covers the case for mutiple axis because if n-1 axis are equal the total prod can only be equal if the 1st are also equal

Reply:
I'm not sure, it sound a bit flaky to me to rely on `dot` handling some runtime check internally. If you want, I can add the change though.

Reply:
Ok, I'll add it

Reply:
Done